### PR TITLE
Issue #696 - Support apps with static namespaces in resources

### PR DESCRIPTION
**Pull request by alexmt**
After PR is merged application might have resources with 'hardcoded' namespaces. This partially solved issue #696

### CODE REVIEW

**Comment by jessesuen** [comment]
I haven't yet looked at the code, but before I forget, do we prevent the following attack?

In git, I have a ClusterRole manifest, with namespace value (even though ClusterRoles don't need namespace). The namespace is in the whitelisted namespaces in the project. ClusterRoles are not whitelisted in the project.

The expected behavior is that the create ClusterRole should be rejected.

**Comment by alexmt** [comment]
@jessesuen thanks for reminding about project permissions check. The attack which you described is prevented ( we are explicetly if resource is namespaced or not using discovery api ). 

I forgot to add destination namespace check. This is fixed now. Please review PR

File: controller/cache/cluster.go [hunk 1]
```diff
@@ -115,6 +116,13 @@ func (c *clusterInfo) getChildren(obj *unstructured.Unstructured) []appv1.Resour
 	return children
 }
 
+func (c *clusterInfo) isNamespaced(gvk schema.GroupVersionKind) bool {
+	if api, ok := c.apis[gvk]; ok && !api.Namespaced {
+		return false
+	}
+	return true
```
Comment (jessesuen) [inline controller/cache/cluster.go]: Is there any problem here that if we don't see the GVK in the cluster, that we unconditionally consider it as a namespaced? e.g. will deploying an app for the first time that contains a cluster-scoped CRD + cluster-scoped CRD object work okay?

**Review by jessesuen — APPROVED** [review]

File: controller/cache/cluster.go (hunk 1 above)
Comment (alexmt) [reply controller/cache/cluster.go]: good point. working improving it in next PR

**Comment by peterbosalliandercom** [comment]
Should this be working on role and rolebindings? I have set a namespace hardcoded into the resource and still it pickes up the application namespace...

**Comment by peterbosalliandercom** [comment]
I have added the following pod, it should be running in namespace demo1 but it runs in demo (the application destination namespace).
apiVersion: v1
kind: Pod
metadata:
  name: myapp-pod
  labels:
    app: myapp
spec:
  containers:
  - name: myapp-container
    image: busybox
    command: ['sh', '-c', 'echo Hello Kubernetes! && sleep 3600']

**Comment by peterbosalliandercom** [comment]
My source pod.yaml was:
apiVersion: v1
kind: Pod
metadata:
  name: myapp-pod
  labels:
    app: myapp
  namespace: demo1
spec:
  containers:
  - name: myapp-container
    image: busybox
    command: ['sh', '-c', 'echo Hello Kubernetes! && sleep 3600']

**Comment by peterbosalliandercom** [comment]
Is this the same issue as described in my previous posts? https://github.com/argoproj/argo-cd/commit/483872a190f39a5520d02fb8bb37a7d8f7b7b4ee

**Comment by jessesuen** [comment]
Yes possibly. Are you using kustomize?

**Pull request merged**